docs/architecture/blazor-for-web-forms-developers/components.md

```diff
@@ -3,11 +3,531 @@ title: Build reusable UI components with Blazor
+--- | --- | ---
+`@attributes` | Render a dictionary of attributes. | `<input @attributes="ExtraAttributes" />`
+`@bind` | Create a two-way data binding. | `<input @bind="username" @bind:event="oninput" />`
+`@on{event}` | Add an event handler for the specified event | `<button @onclick="IncrementCount">Click me!</button>`
+`@key` | Specify a key to be used by the diffing algorithm for preserving elements in a collection | `<DetailsEditor @key="person" Details="person.Details" />`
+`@ref` | Capture a reference to the component or HTML element | `<MyDialog @ref="myDialog" />`
+
+The various directive attributes used by Blazor (`@onclick`, `@bind`, `@ref`, etc.) are covered in the sections below and later chapters.
+
+Many of the syntaxes used in *.aspx* and *.ascx* files have parallel syntaxes in Razor. Below is a simple comparison of the syntaxes for ASP.NET Web Forms and Razor.
+
+Syntax | .aspx | Example | .razor | Example
+--- | --- | --- | --- | ---
+Directives | `<%@ [directive] %>` | `<%@ Page %>` | `@[directive]` | `@page`
+Code blocks | `<% %>` | `<% int x = 123; %>`  | `@{ }` | `@{ int x = 123; }`
+Expressions<br>(HTML encoded) | `<%: %>` | `<%:DateTime.Now %>` | Implicit: `@`<br>Explicit: `@()` | `@DateTime.Now`<br>`@(DateTime.Now)`
+Comments | `<%-- --%>` | `<%-- Commented --%>` | `@* *@` | `@* Commented *@`
+Data binding | `<%# %>` | `<%# Bind("Name") %>` | `@bind` | `<input @bind="username" />`
+
+To add members to the Razor component class use the `@code` directive. This is similar to using a `<script runat="server">...</script>` block in an ASP.NET Web Forms user control or page.
+
+```razor
+@code {
+    int count = 0;
+
+    void IncrementCount() 
+    {
+        count++;
+    }
+}
+```
+
+Note that because Razor is based on C#, it must be compiled from within a C# project (*.csproj*). You cannot compile *.razor* files from a VB project (*.vbproj*). You can still reference VB projects from your Blazor project and vice versa.
+
+A full [Razor syntax reference](https://docs.microsoft.com/aspnet/core/mvc/views/razor) is available in the ASP.NET Core documentation.
+
+## Using components
+
+In addition to normal HTML, components can also use other components as part of their rendering logic. The syntax for using a component in Razor is very similar to using a user control in an ASP.NET Web Forms app. Components are specified using a element tag that matches the type name of the component. For example, you can add a `Counter` component like this:
+
+```razor
+<Counter />
+```
+
+Unlike ASP.NET Web Forms, components in Blazor don't have an element prefix (e.g. "asp:") and they don't need to be registered on the page or in *web.config*. You can think of Razor components like you would .NET types, because that's exactly what they are. If the assembly containing the component is referenced, then the component is available to be used. To bring the component's namespace into scope, you use the `@using` directive.
```

Comment:
```suggestion
Unlike ASP.NET Web Forms, components in Blazor don't have an element prefix (for example, "asp:") and they don't need to be registered on the page or in *web.config*. You can think of Razor components like you would of .NET types, because that's exactly what they are. If the assembly containing the component is referenced, then the component is available to be used. To bring the component's namespace into scope, you use the `@using` directive.
```